### PR TITLE
Fs 3024 db schema - Flags

### PR DESCRIPTION
**Pull request by srh-sloan**
### Change description
Adding DB schema support for the new flags design, as per the prototype: https://fsd-pre-award.herokuapp.com/s39-assessment/flag-enhancements
Allows for flagging updates, history, allocation, and multiple flags.

I've left the existing flag functionality unchanged for now. All new flagging stuff has `v2` in the name to differentiate.

- [x] Unit tests and other appropriate tests added or updated
- [ ] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
Use the swagger interface on assessment-store and the flags_v2 tag

### CODE REVIEW

File: db/models/flags_v2/flag_update.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+from sqlalchemy.sql import func
+
+BaseModel: DefaultMeta = db.Model
+
+
+class FlagStatus(IntEnum):
+    RAISED = 0
+    STOPPED = 1
+    RESOLVED = 3
+
```
Comment (RamuniN) [inline db/models/flags_v2/flag_update.py:20]: Do you need `ESCALATED` type here?

File: openapi/api.yml [hunk 2]
```diff
@@ -678,3 +678,98 @@ paths:
             items:
               type: string
         required: true
+  '/flags_v2/{application_id}':
```
Comment (RamuniN) [inline openapi/api.yml:681]: ok we have `flag_v2` endpoint now. Should we modify assessment flag endpoints to these?

File: db/models/flags_v2/flag_update.py (hunk 1 above)
Comment (srh-sloan) [reply db/models/flags_v2/flag_update.py:20]: Charles said to leave ascalated for now as they weren't sure quite what that meant or if they wanted to ahve it.

File: openapi/api.yml (hunk 2 above)
Comment (srh-sloan) [reply openapi/api.yml:681]: Yes to use the new flag stuff you'll need to use flags_v2. Have left the old stuff there for now so stuff doesn't fallover too much during testing and so we can decide on a migration.,

File: db/migrations/versions/6ad072a15e50_.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            ["assessment_flag.id"],
+            name=op.f("fk_flag_update_assessment_flag_id_assessment_flag"),
+        ),
+        sa.PrimaryKeyConstraint("id", name=op.f("pk_flag_update")),
+    )
+
+    # ### end Alembic commands ###
+
+
+def downgrade():
```
Comment (RamuniN) [inline db/migrations/versions/6ad072a15e50_.py:81]: getting these errors when I run downgrade command `flask db downgrade 14bff16bccc8`. Can you please check this?
```
INFO  [alembic.runtime.migration] Running downgrade 6ad072a15e50 -> 14bff16bccc8, empty message
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.DuplicateObject: extension "pg_trgm" already exists
```

**Comment by RamuniN** [comment]
looks good to me overall. Will need to test this with assessment frontend tomorrow.

File: db/migrations/versions/6ad072a15e50_.py (hunk 3 above)
Comment (srh-sloan) [reply db/migrations/versions/6ad072a15e50_.py:80]: Good spot, sorted now 👍

File: db/models/flags_v2/assessment_flag.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+from sqlalchemy import String
+from sqlalchemy.dialects.postgresql import ARRAY
+from sqlalchemy.dialects.postgresql import ENUM
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.orm import relationship
+
+BaseModel: DefaultMeta = db.Model
+
+
+class AssessmentFlag(BaseModel):
```
Comment (Adam-W1) [inline db/models/flags_v2/assessment_flag.py:17]: might we be able to use a dataclass here with a post init: https://www.infoworld.com/article/3563878/how-to-use-python-dataclasses.html#:~:text=other%20instance%20data%3A

File: api/routes/assessment_routes.py [hunk 5]
```diff
@@ -220,3 +225,26 @@ def assessment_stats_for_fund_round_id(
 
 def get_application_json(application_id):
     return get_application_jsonb_blob(application_id)
+
+
+def get_all_flags_v2_for_application(application_id):
+    current_app.logger.info(f"Get all flags for application {application_id}")
+    flags = get_flags_for_application(application_id)
+    flag_schema = AssessmentFlagSchema()
+    return flag_schema._serialize(flags, many=True)
```
Comment (tferns) [inline api/routes/assessment_routes.py:234]: This looks like we're using an internet `_serialize` method?

Maybe it's better for us to use the public method?

https://github.com/marshmallow-code/marshmallow/blob/5dc855f61759446b3140e3c3bada5a035eab2e1f/src/marshmallow/fields.py#L319-L344

File: db/models/flags_v2/assessment_flag.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        updates,
+        **kwargs,
+    ):
+        self.application_id = application_id
+        self.latest_allocation = latest_allocation
+        self.latest_status = latest_status
+        self.sections_to_flag = sections_to_flag
+        self.updates = updates
+        if "id" in kwargs:
+            self.id = kwargs["id"]
```
Comment (tferns) [inline db/models/flags_v2/assessment_flag.py:33]: Do we need to call the super `__init__` too at all? Or does this not cause any issues as far as you know?

File: tests/conftest.py [hunk 7]
```diff
@@ -48,17 +50,38 @@ def seed_application_records(
...
             app["round_id"] = random_round_id
         app_flags = []
+        app_flags_v2 = []
         if "flags" in app:
             app_flags = app["flags"]
             app.pop("flags")
+        app_flags_v2 = []
+        if "flags_v2" in app:
+            app_flags_v2 = app["flags_v2"]
+            app.pop("flags_v2")
```
Comment (tferns) [inline tests/conftest.py:60]: `pop` looks to return the value so could maybe do this in one line
```suggestion
            app_flags_v2 = app.pop("flags_v2")
```

File: tests/conftest.py [hunk 8]
```diff
@@ -48,17 +50,38 @@ def seed_application_records(
...
         inserted_application = bulk_insert_application_record(
             [app], "COF", True
         )[0]
         app["flags"] = app_flags
+        app["flags_v2"] = app_flags_v2
```
Comment (tferns) [inline tests/conftest.py:65]: Wait I'm confused, why did we `pop` it if we re-add it? 😅

Is it so it doesn't exist when calling `bulk_insert_application_record`?

Oh looking now you've just copied the v1 flag stuff

File: api/routes/assessment_routes.py (hunk 5 above)
Comment (srh-sloan) [reply api/routes/assessment_routes.py:234]: Thanks - it should be `.dump` anyway. Fixed now.

File: db/models/flags_v2/assessment_flag.py (hunk 6 above)
Comment (srh-sloan) [reply db/models/flags_v2/assessment_flag.py:33]: This all works as-is at the moment. I've created a tech debt ticket to start using the sqlalchemy dataclass integration when we upgrade ot sqlalchemy 2.0 as per Adam's comment above.

File: tests/conftest.py (hunk 7 above)
Comment (srh-sloan) [reply tests/conftest.py:60]: Done 👍

File: tests/conftest.py (hunk 8 above)
Comment (srh-sloan) [reply tests/conftest.py:65]: Yeah - we don't want these extra bits of config that we include for seeding test data to get passed to `bulk_insert_application_record` so we take them out and then re-add.

**Review by tferns — APPROVED** [review]

**Comment by tferns** [comment]
Nice work on this! Seems like a big chunk of flagging 👍

**Pull request merged**